source/extensions/filters/http/lua/lua_filter.cc

```diff
@@ -278,7 +278,10 @@ Http::FilterDataStatus StreamHandleWrapper::onData(Buffer::Instance& data, bool
     resumeCoroutine(0, yield_callback_);
   }
 
-  if (state_ == State::HttpCall || state_ == State::WaitForBody) {
+  if (state_ == State::HttpCall) {
+    return filter_.config()->bufferBody() ? Http::FilterDataStatus::StopIterationAndBuffer
```

Comment:
Is this correct? Both `StopIterationAndBuffer` and `StopIterationAndWatermark` will buffer the body. 

The difference here is that former one will return 413 like hard limit and the later on will invoke watermark callbacks `onDecoderFilterAboveWriteBufferHighWatermark` for example.

Reply:
I think it is correct, unless I am missing something. My expectation is that watermark callbacks will indicate to the listener to stop reading from input, therefore putting backpressure on the sender. The use case here is to read headers, issue httpCall (which may be take some time to complete) and pass the body to upstream receiver without investigating it. The problem is that while Lua filter waits for httpCall to complete, client sends so much data that it exceeds configured maximum and connection is terminated.

Reply:
My point is your check here is if  body is buffered, while `StopIterationAndBuffer ` and `StopIterationAndWatermark ` will buffer the body by FM. I think my point is kind of similar to @wbpcode 's [API review](https://github.com/envoyproxy/envoy/pull/35104#discussion_r1669561250)

The difference between those two is not about buffer body instead it is about how overflow data is handled. StopIterationAndBuffer will result in 413 local reply and StopIterationAndWatermark will result in backpressure by watermark callbacks.

Reply:
Got it. Thanks. Yeah, the naming may be misleading. When I was going through the code I realized that Lua filter always buffers body, even when the script does not need the body at all, so I was thinking that having a config option to skip waiting for entire body may be useful for httpCall situation and when script does not need the body (possibly future PR). Hence the name `buffer_body`. But I do not have a full context, so decided to limit the change only for httpCall.

Reply:
>  My expectation is that watermark callbacks will indicate to the listener to stop reading from input, therefore putting backpressure on the sender.

Is this ensured that Envoy will do that? If yes then please comment it in the code to let the future developer/reader know that.